xc/device.ml

```diff
@@ -1984,15 +1979,36 @@ module Backend = struct
           let qmp_event_handle domid qmp_event =
             (* This function will be extended to handle qmp events *)
             debug "Got QMP event, domain-%d: %s" domid qmp_event.event;
-            qmp_event.data >>= function
-            | RTC_CHANGE timeoffset ->
+
+            let rtc_change timeoffset =
               with_xs (fun xs ->
                 let timeoffset_key = sprintf "/vm/%s/rtc/timeoffset" (Uuidm.to_string (Xenops_helpers.uuid_of_domid ~xs domid)) in
                 try
                   let rtc = xs.Xs.read timeoffset_key in
                   xs.Xs.write timeoffset_key Int64.(add timeoffset (of_string rtc) |> to_string)
                 with e -> error "Failed to process RTC_CHANGE for domain %d: %s" domid (Printexc.to_string e)
               )
+            in
+
+            let xen_platform_pv_driver_info pv_info =
+              with_xs (fun xs ->
+                let is_hvm_linux { product_num; build_num } =
+                  let _XEN_IOPORT_LINUX_PRODNUM = 3 in (* from Linux include/xen/platform_pci.h *)
+                  (product_num = _XEN_IOPORT_LINUX_PRODNUM) && (build_num <= 0xff)
+                in
+                if is_hvm_linux pv_info then
+                begin
+                  let write_local_domain prefix x = xs.Xs.write (Printf.sprintf "/local/domain/%d/%s%s" domid prefix x) "1" in
+                  List.iter (write_local_domain "control/feature-") ["suspend"; "poweroff"; "reboot"; "vcpu-hotplug"];
+                  List.iter (write_local_domain "data/") ["updated"]
+                end
+              )
+            in
+
+            qmp_event.data |> function
+            | Some (RTC_CHANGE timeoffset)         -> rtc_change timeoffset
+            | Some (XEN_PLATFORM_PV_DRIVER_INFO x) -> xen_platform_pv_driver_info x
+            | _ -> () (* unhandled QMP events *)
```

Comment:
Should this be logged or do you expect this to be a normal case?

Reply:
I think logging the wildcard case will cause spam in the logs, because all incoming events are already logged by line https://github.com/mg12/xenopsd/blob/0615ef98eb56f28705c77cd6e866d38824565565/xc/device.ml#L1981 and I think it's also unnecessary to log the wildcard case because it's possible to easily infer which of the incoming events are not being handled by checking if they are in the whitelist of handled events that currently contains RTC_CHANGE and XEN_PLATFORM_PV_DRIVER_INFO. Does that sound reasonable?